scipy/optimize/_constraints.py

```diff
@@ -133,25 +133,75 @@ class LinearConstraint:
     ----------
     A : {array_like, sparse matrix}, shape (m, n)
         Matrix defining the constraint.
-    lb, ub : array_like
-        Lower and upper bounds on the constraint. Each array must have the
+    lb, ub : array_like, optional
+        Lower and upper limits on the constraint. Each array must have the
         shape (m,) or be a scalar, in the latter case a bound will be the same
         for all components of the constraint. Use ``np.inf`` with an
         appropriate sign to specify a one-sided constraint.
         Set components of `lb` and `ub` equal to represent an equality
         constraint. Note that you can mix constraints of different types:
         interval, one-sided or equality, by setting different components of
-        `lb` and `ub` as  necessary.
+        `lb` and `ub` as  necessary. Defaults to ``lb = -np.inf``
+        and ``ub = np.inf`` (no limits).
     keep_feasible : array_like of bool, optional
         Whether to keep the constraint components feasible throughout
         iterations. A single value set this property for all components.
         Default is False. Has no effect for equality constraints.
     """
-    def __init__(self, A, lb, ub, keep_feasible=False):
-        self.A = A
-        self.lb = lb
-        self.ub = ub
-        self.keep_feasible = keep_feasible
+    def _input_validation(self):
+        if self.A.ndim != 2:
+            message = "`A` must have exactly two dimensions."
+            raise ValueError(message)
+
+        try:
+            shape = self.A.shape[0:1]
```

Comment:
`self.A.shape[0]` seems clearer to me -- any reason to prefer `[0:1]`?  (here and on line 163)

Reply:
It retains the parentheses.

e.g. 
```python3
>>> A.shape[0]
2

>>> A.shape[0:1]
(2,)
```
which is appropriate because shapes are supposed to be tuples, not scalars (even if they have just one element).
```
>>> np.zeros(2).shape
(2,)
```

That said, I can't remember why I did it, because I suspect that's not the original reason : P `broadcast_to` doesn't mind if `shape` is a scalar; it turns it into a tuple on its own. Since `A` is known to be 2D at this point, `shape` will definitely have two elements, so it's not because `()[0]` is an error and `()[0:1]` returns `()`. I'd hate to remove it and find a bug, though.

Reply:
I didn't realize it was a single-element tuple -- learn something new everyday!